src/learning/agglomerative.rs

```diff
@@ -0,0 +1,415 @@
+            },
+            &Metrics::Ward1  => {
+                let s = ci.size + cj.size + ck.size;
+                let dij = dmat.get(ci.id, cj.id);
+                (ci.size + ck.size) / s * dik + (cj.size + ck.size) / s * djk - ck.size / s * dij
+            },
+            &Metrics::Ward | &Metrics::Ward2 => {
+                let s = ci.size + cj.size + ck.size;
+                let dij = dmat.get(ci.id, cj.id);
+                ((ci.size + ck.size) / s * dik.powf(2.) + (cj.size + ck.size) / s * djk.powf(2.) - ck.size / s * dij.powf(2.)).sqrt()
+            }
+        }
+    }
+}
+
+struct Cluster {
+    /// Cluster id
+    id: usize,
+    /// Number of nodes (rows) which belongs to cluster
+    /// to avoid cast in the algorithm, store it as f64
+    size: f64,
+    /// Row ids belong to the cluster
+    nodes: Vec<usize>,
+}
+
+impl Cluster {
+
+    /// Create new cluster
+    fn new(id: usize, nodes: Vec<usize>) -> Cluster {
+        Cluster {
+            id: id,
+            size: 1.,
+            nodes: nodes,
+        }
+    }
+
+    /// Create new cluster merging left and right
+    fn from_clusters(id: usize, left: Cluster, mut right: Cluster) -> Cluster {
+        let mut new_nodes = left.nodes;
+        new_nodes.append(&mut right.nodes);
+        Cluster {
+            id: id,
+            size: left.size + right.size,
+            nodes: new_nodes
+        }
+    }
+}
+
+/// Distance Matrix
+#[derive(Debug)]
+struct DistanceMatrix {
+    // Distance is symmetric, no need to hold all pairs
+    // use HashMap to easier update
+    data: HashMap<(usize, usize), f64>
+}
+
+impl DistanceMatrix {
+
+    /// Create distance matrix fron input matrix
+    fn from_mat(inputs: &Matrix<f64>) -> Self {
+        assert!(inputs.rows() > 0, "input is empty");
+
+        let n = inputs.rows() - 1;
+        let mut data: HashMap<(usize, usize), f64> = HashMap::with_capacity(n * n);
+
+        unsafe {
+            for i in 0..n {
+                for j in i..inputs.rows() {
+                    let mut val = 0.;
+                    for k in 0..inputs.cols() {
+                        val += (inputs.get_unchecked([i, k]) -
+                                inputs.get_unchecked([j, k])).abs().powf(2.);
+                    }
+                    val = val.sqrt();
+                    data.insert((i, j), val);
+                }
+            }
+        }
+        DistanceMatrix {
+            data: data
+        }
+    }
+
+    /// Get distance between i-th and j-th item
+    fn get(&self, i: usize, j: usize) -> f64 {
+        if i == j {
+            0.
+        } else if i > j {
+            *self.data.get(&(j, i)).unwrap()
+        } else {
+            *self.data.get(&(i, j)).unwrap()
+        }
+    }
+
+    /// Add distance between i-th and j-th item
+    /// i must be smaller than j
+    fn insert(&mut self, i: usize, j: usize, dist: f64) {
+        assert!(i < j, "i must be smaller than j");
```

Comment:
I would make this `assert_debug`. The function is only used internally so we wouldn't expect a failure - as such we don't want a performance hit when run in release mode.